pkg/authorization/admission/restrictusers/restrictusers.go

```diff
@@ -0,0 +1,168 @@
+// subjects.
+type restrictUsersAdmission struct {
+	*admission.Handler
+
+	oclient oclient.Interface
+	kclient kclientset.Interface
+}
+
+var _ = oadmission.WantsOpenshiftClient(&restrictUsersAdmission{})
+var _ = oadmission.Validator(&restrictUsersAdmission{})
+
+// NewRestrictUsersAdmission configures an admission plugin that enforces
+// restrictions on adding role bindings in a project.
+func NewRestrictUsersAdmission(kclient kclientset.Interface) (admission.Interface, error) {
+	return &restrictUsersAdmission{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+		kclient: kclient,
+	}, nil
+}
+
+func (q *restrictUsersAdmission) SetOpenshiftClient(c oclient.Interface) {
+	q.oclient = c
+}
+
+// rolebindingsDelta returns the relative complement of rolebindings a in
+// rolebindings b (i.e., a∖b).
+func rolebindingsDelta(a, b *authorizationapi.RoleBinding) []kapi.ObjectReference {
+	result := []kapi.ObjectReference{}
+
+	for _, y := range b.Subjects {
+		for _, x := range a.Subjects {
+			if y == x {
+				goto nexty
+			}
+		}
+		result = append(result, y)
+	nexty:
+	}
+
+	return result
+}
+
+// Admit makes admission decisions that enforce restrictions on adding
+// project-scoped role-bindings.  In order for a role binding to be permitted,
+// each subject in the binding must be matched by some rolebinding restriction
+// in the namespace.
+func (q *restrictUsersAdmission) Admit(a admission.Attributes) (err error) {
+	if a.GetResource().GroupResource() != authorizationapi.Resource("rolebindings") {
+		return nil
+	}
+
+	// Ignore all operations that correspond to subresource actions.
+	if len(a.GetSubresource()) != 0 {
+		return nil
+	}
+
+	ns := a.GetNamespace()
+
+	// Ignore cluster-level resources.
+	if len(ns) == 0 {
+		return nil
+	}
+
+	binding, ok := a.GetObject().(*authorizationapi.RoleBinding)
+	if !ok {
+		return admission.NewForbidden(a,
+			fmt.Errorf("wrong object type for new rolebinding: %t", a.GetObject()))
+	}
+
+	oldbinding, ok := a.GetOldObject().(*authorizationapi.RoleBinding)
+	if oldbinding != nil && !ok {
+		return admission.NewForbidden(a,
+			fmt.Errorf("wrong object type for old rolebinding: %t", a.GetOldObject()))
+	}
+
+	glog.V(4).Infof("Handling rolebinding %s/%s", binding.Namespace, binding.Name)
+
+	var newSubjects []kapi.ObjectReference
+
+	if oldbinding == nil {
+		newSubjects = binding.Subjects
+	} else {
+		newSubjects = rolebindingsDelta(oldbinding, binding)
+	}
+
+	if len(newSubjects) == 0 {
+		glog.V(4).Infof("No new subjects; admitting")
+		return nil
+	}
+
+	// TODO: Cache rolebinding restrictions.
+	matchersList, err := q.oclient.RoleBindingRestrictions(ns).List(kapi.ListOptions{})
+	if err != nil {
+		return admission.NewForbidden(a, err)
+	}
+	if len(matchersList.Items) == 0 {
+		glog.V(4).Infof("No rolebinding restrictions specified; admitting")
+		return nil
+	}
+
+	rbrCtx, err := newRoleBindingRestrictionContext(ns, q.kclient, q.oclient)
```

Comment:
At this point, build the slice of `SubjectCheckers`, then construct the `UnionSubjectChecker` (I think that's just `type UnionSubjectChecker []SubjectCheckers`, then run the check with the context.